--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -516,6 +_,7 @@
 
         @PostProcess
         private void postProcess() {
+            this.preventMovingIntoUnloadedChunks = true; // Canvas - prevent moving into unloaded chunks
             FeatureHooks.setPlayerChunkUnloadDelay(this.delayChunkUnloadsBy.ticks());
         }
     }
