--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityType.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityType.java
@@ -1,6 +_,7 @@
 package org.bukkit.craftbukkit.entity;
 
 import com.google.common.base.Preconditions;
+import io.canvasmc.canvas.Config;
 import io.papermc.paper.registry.RegistryKey;
 import java.util.Locale;
 import net.minecraft.core.Holder;
@@ -14,14 +_,23 @@
 import org.bukkit.entity.EntityType;
 
 public class CraftEntityType {
+    private static final java.util.Map<net.minecraft.world.entity.EntityType<?>, EntityType> MINECRAFT_TO_BUKKIT_KEY_CACHE = new java.util.concurrent.ConcurrentHashMap<>(); // Canvas - cache CraftEntityType#minecraftToBukkit convert
 
     public static EntityType minecraftToBukkit(net.minecraft.world.entity.EntityType<?> minecraft) {
         Preconditions.checkArgument(minecraft != null);
+        // Canvas start - cache CraftEntityType#minecraftToBukkit convert
+        EntityType asBukkit = MINECRAFT_TO_BUKKIT_KEY_CACHE.get(minecraft);
+
+        if (Config.INSTANCE.entities.cacheMinecraftToBukkitEntityTypeConvert && asBukkit != null) {
+            return asBukkit;
+        }
+        // Canvas end
 
         net.minecraft.core.Registry<net.minecraft.world.entity.EntityType<?>> registry = CraftRegistry.getMinecraftRegistry(Registries.ENTITY_TYPE);
         EntityType bukkit = Registry.ENTITY_TYPE.get(CraftNamespacedKey.fromMinecraft(registry.getResourceKey(minecraft).orElseThrow().location()));
 
         Preconditions.checkArgument(bukkit != null);
+        MINECRAFT_TO_BUKKIT_KEY_CACHE.put(minecraft, bukkit); // Canvas - cache CraftEntityType#minecraftToBukkit convert
 
         return bukkit;
     }
