--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -6744,6 +_,7 @@
             .mapColor(MapColor.COLOR_ORANGE)
             .instrument(NoteBlockInstrument.BASEDRUM)
             .requiresCorrectToolForDrops()
+            .sound(SoundType.COPPER) // Canvas - MC-223153
             .strength(5.0F, 6.0F)
     );
     public static final Block RAW_GOLD_BLOCK = register(
@@ -7077,4 +_,13 @@
             }
         }
     }
+    // Canvas start - cache path type
+    public static void initPathType() {
+        for (Block block : BuiltInRegistries.BLOCK) {
+            for (BlockState blockState : block.getStateDefinition().getPossibleStates()) {
+                blockState.initPathType();
+            }
+        }
+    }
+    // Canvas end
 }
