--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -29,6 +_,7 @@
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
+import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.material.FlowingFluid;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.FluidState;
@@ -189,17 +_,22 @@
 
             for (Direction direction : POSSIBLE_FLOW_DIRECTIONS) {
                 BlockPos blockPos = pos.relative(direction.getOpposite());
-                if (level.getFluidState(blockPos).is(FluidTags.WATER)) {
+                // Canvas start - don't load chunks to spread liquid
+                LevelChunk chunk = level.getChunkIfLoaded(blockPos);
+                if (chunk == null) continue; // chunk not loaded, skip.
+                if (chunk.getFluidState(blockPos).is(FluidTags.WATER)) {
+                // Canvas end
                     Block block = level.getFluidState(pos).isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
                     // CraftBukkit start
                     if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, block.defaultBlockState(), 3)) {
+                        level.setBlock(pos, block.defaultBlockState(), 3); // Canvas - place block
                         this.fizz(level, pos);
                     }
                     // CraftBukkit end
                     return false;
                 }
 
-                if (isSoulSoil && level.getBlockState(blockPos).is(Blocks.BLUE_ICE)) {
+                if (isSoulSoil && chunk.getBlockState(blockPos).is(Blocks.BLUE_ICE)) { // Canvas - don't load chunks to spread liquid
                     // CraftBukkit start
                     if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, Blocks.BASALT.defaultBlockState(), 3)) {
                         this.fizz(level, pos);
