--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2534,4 +_,11 @@
         return POINTERS_SUPPLIER.view(this);
     }
     // Paper end
+    // Canvas start - expand region api
+
+    @Override
+    public io.canvasmc.canvas.threadedregions.ServerRegionizer getRegionizer() {
+        return this.getHandle().regioniser.apiRegionizer;
+    }
+    // Canvas end
 }
