--- a/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
+++ b/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
@@ -185,9 +_,15 @@
             return;
         }
 
-        if (!Bukkit.isPrimaryThread()) {
+        if (!Bukkit.isPrimaryThread() && !(Thread.currentThread() instanceof io.canvasmc.canvas.server.chunk.AsyncChunkSend.AsyncChunkSendThread)) { // Canvas - async chunk sending
             // Plugins?
-            MinecraftServer.getServer().scheduleOnMain(() -> modifyBlocks(chunkPacket, chunkPacketInfo));
+            // Canvas start - Threaded Regions
+            int chunkX = chunkPacket.getX();
+            int chunkZ = chunkPacket.getZ();
+            MinecraftServer.getThreadedServer().taskQueue.queueTickTaskQueue(
+                chunkPacketInfo.getChunk().level, chunkX, chunkZ, () -> modifyBlocks(chunkPacket, chunkPacketInfo)
+            );
+            // Canvas end
             return;
         }
 
