--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -232,6 +_,12 @@
 
     @Nullable
     public Vec3 getPortalPosition(ServerLevel level, BlockPos pos) {
+        // Canvas start - ensure tick thread
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(level, pos)) {
+            this.trySearchForExit(level, pos);
+            return null;
+        }
+        // Canvas end
         if (this.exitPortal == null && level.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.END) { // CraftBukkit - work in alternate worlds
             BlockPos blockPos = findOrCreateValidTeleportPos(level, pos);
             blockPos = blockPos.above(10);
