--- a/net/minecraft/world/entity/vehicle/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/AbstractMinecart.java
@@ -492,6 +_,7 @@
     public void applyEffectsFromBlocks() {
         if (!useExperimentalMovement(this.level())) {
             this.applyEffectsFromBlocks(this.position(), this.position());
+            this.clearMovementsThisTick(); // Canvas - temporary memory leak fix
         } else {
             super.applyEffectsFromBlocks();
         }
