--- a/net/minecraft/network/protocol/PacketUtils.java
+++ b/net/minecraft/network/protocol/PacketUtils.java
@@ -50,7 +_,7 @@
             };
             // ignore retired state, if removed then we don't want the packet to be handled
             if (processor instanceof net.minecraft.server.network.ServerGamePacketListenerImpl gamePacketListener) {
-                gamePacketListener.player.queuePacketTask(run);
+                gamePacketListener.player.getBukkitEntity().addPacket(run); // Canvas - rewrite scheduler
             } else if (processor instanceof net.minecraft.server.network.ServerConfigurationPacketListenerImpl configurationPacketListener) {
                 io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTask(run);
             } else if (processor instanceof net.minecraft.server.network.ServerLoginPacketListenerImpl loginPacketListener) {
