--- a/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
+++ b/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
@@ -160,6 +_,7 @@
         }
 
         this.vanillaRender.buffer = colors.array(); // Paper - Use Vanilla map renderer when possible
+        this.setDirty(false); // Canvas - reset dirty flag when loading maps from disk
     }
 
     public static MapItemSavedData createFresh(
