--- a/src/main/java/ca/spottedleaf/moonrise/common/util/CoordinateUtils.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/CoordinateUtils.java
@@ -21,7 +_,7 @@
     }
 
     public static long getChunkKey(final ChunkPos pos) {
-        return ((long)pos.z << 32) | (pos.x & 0xFFFFFFFFL);
+        return pos.longKey; // Canvas - use pos longKey
     }
 
     public static long getChunkKey(final SectionPos pos) {
