--- a/net/minecraft/world/entity/npc/CatSpawner.java
+++ b/net/minecraft/world/entity/npc/CatSpawner.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.entity.npc;
 
 import java.util.List;
+import io.canvasmc.canvas.region.ServerRegions;
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.tags.StructureTags;
@@ -18,15 +_,16 @@
 
 public class CatSpawner implements CustomSpawner {
     private static final int TICK_DELAY = 1200;
-    private int nextTick;
+    // private int nextTick; // Canvas - Threaded Regions
 
     @Override
     public void tick(ServerLevel level, boolean spawnEnemies, boolean spawnFriendlies) {
         if (spawnFriendlies && level.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
-            this.nextTick--;
-            if (this.nextTick <= 0) {
-                this.nextTick = level.purpurConfig.catSpawnDelay; // Purpur - Cat spawning options
-                Player randomPlayer = level.getRandomPlayer();
+            ServerRegions.WorldTickData tickData = ServerRegions.getTickData(level); // Canvas - Threaded Regions
+            tickData.catSpawnerNextTick--; // Canvas - Threaded Regions
+            if (tickData.catSpawnerNextTick <= 0) { // Canvas - Threaded Regions
+                tickData.catSpawnerNextTick = level.purpurConfig.catSpawnDelay; // Purpur - Cat spawning options // Canvas - Threaded Regions
+                Player randomPlayer = level.getRandomLocalPlayer(); // Canvas - Threaded Regions
                 if (randomPlayer != null) {
                     RandomSource randomSource = level.random;
                     int i = (8 + randomSource.nextInt(24)) * (randomSource.nextBoolean() ? -1 : 1);
