--- a/net/minecraft/world/level/LevelReader.java
+++ b/net/minecraft/world/level/LevelReader.java
@@ -57,6 +_,12 @@
         return this.getBiomeManager().getBiome(pos);
     }
 
+    // Canvas start - cache getBiome
+    default Holder<Biome> getBiomeCached(@Nullable net.minecraft.world.level.chunk.LevelChunk chunk, BlockPos pos) {
+        return this.getBiomeManager().getBiomeCached(chunk, pos);
+    }
+    // Canvas end - cache getBiome
+
     default Stream<BlockState> getBlockStatesIfLoaded(AABB aabb) {
         int floor = Mth.floor(aabb.minX);
         int floor1 = Mth.floor(aabb.maxX);
