--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -1,11 +_,15 @@
 package net.minecraft.server.level;
 
+import ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel;
 import com.google.common.annotations.VisibleForTesting;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
+import io.canvasmc.canvas.region.ServerRegions;
 import it.unimi.dsi.fastutil.longs.LongSet;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet;
+import io.canvasmc.canvas.Config;
+import io.netty.util.internal.ConcurrentSet;
 import java.io.IOException;
 import java.nio.file.Path;
 import java.util.Arrays;
@@ -57,9 +_,9 @@
 
 public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moonrise.patches.chunk_system.world.ChunkSystemServerChunkCache { // Paper - rewrite chunk system
     private static final Logger LOGGER = LogUtils.getLogger();
-    private final DistanceManager distanceManager;
+    public final DistanceManager distanceManager; // Canvas - private -> public
     private final ServerLevel level;
-    public final Thread mainThread;
+    public Thread mainThread; // Canvas - mutable
     final ThreadedLevelLightEngine lightEngine;
     public final ServerChunkCache.MainThreadExecutor mainThreadProcessor;
     public final ChunkMap chunkMap;
@@ -72,17 +_,18 @@
     private final long[] lastChunkPos = new long[4];
     private final ChunkStatus[] lastChunkStatus = new ChunkStatus[4];
     private final ChunkAccess[] lastChunk = new ChunkAccess[4];
-    private final List<LevelChunk> spawningChunks = new ObjectArrayList<>();
+    /* private final List<LevelChunk> spawningChunks = new ObjectArrayList<>(); // Canvas - Threaded Regions
     private final Set<ChunkHolder> chunkHoldersToBroadcast = new ReferenceOpenHashSet<>();
     @Nullable
     @VisibleForDebug
-    private NaturalSpawner.SpawnState lastSpawnState;
+    private NaturalSpawner.SpawnState lastSpawnState; */ // Canvas - Threaded Regions
     // Paper start
     private final ca.spottedleaf.concurrentutil.map.ConcurrentLong2ReferenceChainedHashTable<net.minecraft.world.level.chunk.LevelChunk> fullChunks = new ca.spottedleaf.concurrentutil.map.ConcurrentLong2ReferenceChainedHashTable<>();
     public int getFullChunksCount() {
         return this.fullChunks.size();
     }
     long chunkFutureAwaitCounter;
+    public int ticksSinceLastPurgeStaleTicketsCall = 0; // Canvas
     // Paper end
     // Paper start - rewrite chunk system
 
@@ -138,6 +_,14 @@
         final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder currentChunk = chunkHolderManager.getChunkHolder(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ));
 
         final ChunkAccess ifPresent = currentChunk == null ? null : currentChunk.getChunkIfPresent(toStatus);
+        // Canvas start - try prevent syncload
+        if (ifPresent == null) {
+            ChunkAccess attempt = this.fullChunks.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ));
+            if (attempt != null) {
+                return attempt;
+            }
+        }
+        // Canvas end
 
         if (ifPresent != null && (toStatus != ChunkStatus.FULL || currentChunk.isFullChunkReady())) {
             return ifPresent;
@@ -173,11 +_,11 @@
 
         java.util.Objects.checkFromToIndex(0, size, raw.length);
         for (int i = 0; i < size; ++i) {
-            world.tickChunk(raw[i], randomTickSpeed);
+            LevelChunk chunk = raw[i]; if (chunk == null) continue; world.tickChunk(chunk, randomTickSpeed); // Canvas - ensure non null
 
             // call mid-tick tasks for chunk system
             if ((i & 7) == 0) {
-                ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.level.getServer()).moonrise$executeMidTickTasks();
+                this.pollTask(); // Canvas - poll task, as executeMidTickTasks does nothing now
                 continue;
             }
         }
@@ -284,6 +_,12 @@
     public LevelChunk getChunkAtIfLoadedImmediately(int x, int z) {
         return this.fullChunks.get(ChunkPos.asLong(x, z));
     }
+    // Canvas start
+
+    @Nullable
+    public LevelChunk getChunkAtIfLoadedImmediately(long chunkKey) {
+        return this.fullChunks.get(chunkKey);
+    }
     // Paper end
 
     @Override
@@ -330,6 +_,13 @@
         // Paper end - rewrite chunk system
     }
 
+    // Canvas start
+    @Nullable
+    public LevelChunk getChunkNow(ChunkPos pos) {
+        return getChunkNow(pos.x, pos.z);
+    }
+
+    // Canvas end
     @Nullable
     @Override
     public LevelChunk getChunkNow(int chunkX, int chunkZ) {
@@ -446,8 +_,7 @@
 
     public boolean isPositionTicking(long chunkPos) {
         // Paper start - rewrite chunk system
-        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder newChunkHolder = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler().chunkHolderManager.getChunkHolder(chunkPos);
-        return newChunkHolder != null && newChunkHolder.isTickingReady();
+        return ((ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler().chunkHolderManager.blockTickingChunkHolders.contains(chunkPos); // Canvas - faster impl
         // Paper end - rewrite chunk system
     }
 
@@ -489,7 +_,12 @@
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("purge");
         if (this.level.tickRateManager().runsNormally() || !tickChunks || this.level.spigotConfig.unloadFrozenChunks) { // Spigot
+            // Canvas start
+            if (this.ticksSinceLastPurgeStaleTicketsCall++ > Config.INSTANCE.ticksBetweenPurgeStaleTickets) {
             this.ticketStorage.purgeStaleTickets();
+            this.ticksSinceLastPurgeStaleTicketsCall = 0;
+            }
+            // Canvas end
         }
 
         this.runDistanceManagerUpdates();
@@ -508,7 +_,8 @@
 
     private void tickChunks() {
         long gameTime = this.level.getGameTime();
-        long l = gameTime - this.lastInhabitedUpdate;
+        ServerRegions.WorldTickData tickData = ServerRegions.getTickData(this.level); // Canvas - Threaded Regions
+        long l = 1L; // Canvas - Threaded Regions
         this.lastInhabitedUpdate = gameTime;
         if (!this.level.isDebug()) {
             ProfilerFiller profilerFiller = Profiler.get();
@@ -521,31 +_,60 @@
 
             this.broadcastChangedChunks(profilerFiller);
             profilerFiller.pop();
+            // Canvas start - async mob spawning
+            if (Config.INSTANCE.entities.enableAsyncSpawning) {
+                for (ServerPlayer player : tickData.getLocalPlayers()) { // Canvas - Threaded Regions
+                    // Paper start - per player mob spawning backoff
+                    for (int ii = 0; ii < ServerPlayer.MOBCATEGORY_TOTAL_ENUMS; ii++) {
+                        player.mobCounts[ii] = 0;
+                        int newBackoff = player.mobBackoffCounts[ii] - 1; // TODO make configurable bleed // TODO use nonlinear algorithm?
+                        if (newBackoff < 0) {
+                            newBackoff = 0;
+                        }
+                        player.mobBackoffCounts[ii] = newBackoff;
+                    }
+                    // Paper end - per player mob spawning backoff
+               }
+                if (tickData.firstRunSpawnCounts) {
+                    tickData.firstRunSpawnCounts = false;
+                    tickData.spawnCountsReady.set(true);
+                }
+                if (tickData.spawnCountsReady.getAndSet(false)) {
+                    net.minecraft.server.MinecraftServer.getServer().mobSpawnExecutor.submit(() -> {
+                        int mapped = distanceManager.getNaturalSpawnChunkCount();
+                        tickData.setLastSpawnState(NaturalSpawner.createState(mapped, tickData.entityTickList, this::getFullChunk, null, true)); // Canvas - Threaded Regions
+                        tickData.spawnCountsReady.set(true);
+                    });
+                }
+            }
+            // Canvas end
         }
     }
 
     private void broadcastChangedChunks(ProfilerFiller profiler) {
         profiler.push("broadcast");
 
-        for (ChunkHolder chunkHolder : this.chunkHoldersToBroadcast) {
+        for (ChunkHolder chunkHolder : ServerRegions.getTickData(this.level).getChunkHoldersToBroadcast()) { // Canvas - Threaded Regions
             LevelChunk tickingChunk = chunkHolder.getChunkToSend(); // Paper - rewrite chunk system
             if (tickingChunk != null) {
                 chunkHolder.broadcastChanges(tickingChunk);
             }
         }
 
-        this.chunkHoldersToBroadcast.clear();
+        ServerRegions.getTickData(this.level).getChunkHoldersToBroadcast().clear(); // Canvas - Threaded Regions
         profiler.pop();
     }
 
     private void tickChunks(ProfilerFiller profiler, long timeInhabited) {
+        ServerRegions.WorldTickData tickData = ServerRegions.getTickData(this.level); // Canvas - Threaded Regions
         profiler.popPush("naturalSpawnCount");
         int naturalSpawnChunkCount = this.distanceManager.getNaturalSpawnChunkCount();
         // Paper start - Optional per player mob spawns
         NaturalSpawner.SpawnState spawnState;
         if ((this.spawnFriendlies || this.spawnEnemies) && this.level.paperConfig().entities.spawning.perPlayerMobSpawns) { // don't count mobs when animals and monsters are disabled
+            if (!Config.INSTANCE.entities.enableAsyncSpawning) { // Canvas - async mob spawning
             // re-set mob counts
-            for (ServerPlayer player : this.level.players) {
+            for (ServerPlayer player : this.level.getLocalPlayers()) { // Canvas - Threaded Regions
                 // Paper start - per player mob spawning backoff
                 for (int ii = 0; ii < ServerPlayer.MOBCATEGORY_TOTAL_ENUMS; ii++) {
                     player.mobCounts[ii] = 0;
@@ -558,48 +_,53 @@
                 }
                 // Paper end - per player mob spawning backoff
             }
-            spawnState = NaturalSpawner.createState(naturalSpawnChunkCount, this.level.getAllEntities(), this::getFullChunk, null, true);
+            tickData.setLastSpawnState(NaturalSpawner.createState(naturalSpawnChunkCount, this.level.getAllEntities(), this::getFullChunk, null, true)); // Canvas - Threaded Regions
+            } // Canvas - async mob spawning
         } else {
-            spawnState = NaturalSpawner.createState(naturalSpawnChunkCount, this.level.getAllEntities(), this::getFullChunk, !this.level.paperConfig().entities.spawning.perPlayerMobSpawns ? new LocalMobCapCalculator(this.chunkMap) : null, false);
+            tickData.setLastSpawnState(NaturalSpawner.createState(naturalSpawnChunkCount, this.level.getAllEntities(), this::getFullChunk, !this.level.paperConfig().entities.spawning.perPlayerMobSpawns ? new LocalMobCapCalculator(this.chunkMap) : null, false)); // Canvas - Threaded Regions
+            tickData.spawnCountsReady.set(true); // Canvas - async mob spawning
         }
         // Paper end - Optional per player mob spawns
-        this.lastSpawnState = spawnState;
+        // this.lastSpawnState = spawnState; // Canvas - async mob spawning
         profiler.popPush("spawnAndTick");
-        boolean _boolean = this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING) && !this.level.players().isEmpty(); // CraftBukkit
+        boolean _boolean = this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING) && !this.level.getLocalPlayers().isEmpty(); // CraftBukkit // Canvas - Threaded Regions
         int _int = this.level.getGameRules().getInt(GameRules.RULE_RANDOMTICKING);
         List<MobCategory> filteredSpawningCategories;
         if (_boolean && (this.spawnEnemies || this.spawnFriendlies)) {
             // Paper start - PlayerNaturallySpawnCreaturesEvent
-            for (ServerPlayer entityPlayer : this.level.players()) {
+            for (ServerPlayer entityPlayer : tickData.getLocalPlayers()) { // Canvas - Threaded Regions
                 int chunkRange = Math.min(level.spigotConfig.mobSpawnRange, entityPlayer.getBukkitEntity().getViewDistance());
                 chunkRange = Math.min(chunkRange, 8);
                 entityPlayer.playerNaturallySpawnedEvent = new com.destroystokyo.paper.event.entity.PlayerNaturallySpawnCreaturesEvent(entityPlayer.getBukkitEntity(), (byte) chunkRange);
                 entityPlayer.playerNaturallySpawnedEvent.callEvent();
             }
             // Paper end - PlayerNaturallySpawnCreaturesEvent
-            boolean flag = this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) != 0L && this.level.getLevelData().getGameTime() % this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) == 0L; // CraftBukkit
-            filteredSpawningCategories = NaturalSpawner.getFilteredSpawningCategories(spawnState, this.spawnFriendlies, this.spawnEnemies, flag, this.level); // CraftBukkit
+            boolean flag = this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) != 0L && this.level.getRedstoneGameTime() % this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) == 0L; // CraftBukkit // Canvas - Threaded Regions
+            filteredSpawningCategories = NaturalSpawner.getFilteredSpawningCategories(tickData.getLastSpawnState(), this.spawnFriendlies, this.spawnEnemies, flag, this.level); // CraftBukkit // Canvas - Threaded Regions
         } else {
             filteredSpawningCategories = List.of();
         }
 
-        List<LevelChunk> list = this.spawningChunks;
+        // List<LevelChunk> list = this.spawningChunks; // Canvas - optimize chunk collect
 
         try {
             profiler.push("filteringSpawningChunks");
-            this.chunkMap.collectSpawningChunks(list);
+            LevelChunk[] chunks = this.chunkMap.collectSpawningChunks(tickData);
             profiler.popPush("shuffleSpawningChunks");
             // Paper start - chunk tick iteration optimisation
-            this.shuffleRandom.setSeed(this.level.random.nextLong());
-            if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) Util.shuffle(list, this.shuffleRandom); // Paper - Optional per player mob spawns; do not need this when per-player is enabled
+            if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) {
+                this.shuffleRandom.setSeed(this.level.random.nextLong()); // Canvas - only set seed if is really used
+                // Canvas - optimize chunk collection
+            }
             // Paper end - chunk tick iteration optimisation
             profiler.popPush("tickSpawningChunks");
 
-            for (LevelChunk levelChunk : list) {
-                this.tickSpawningChunk(levelChunk, timeInhabited, filteredSpawningCategories, spawnState);
+            for (LevelChunk levelChunk : chunks) { // Canvas - optimize chunk collect
+                if (levelChunk == null) continue; // Canvas - optimize chunk collect
+                this.tickSpawningChunk(levelChunk, timeInhabited, filteredSpawningCategories, tickData.getLastSpawnState(), tickData); // Canvas - Threaded Regions
             }
         } finally {
-            list.clear();
+            // list.clear(); // Canvas - optimize chunk collect
         }
 
         profiler.popPush("tickTickingChunks");
@@ -611,7 +_,7 @@
         }
     }
 
-    private void tickSpawningChunk(LevelChunk chunk, long timeInhabited, List<MobCategory> spawnCategories, NaturalSpawner.SpawnState spawnState) {
+    private void tickSpawningChunk(LevelChunk chunk, long timeInhabited, List<MobCategory> spawnCategories, NaturalSpawner.SpawnState spawnState, ServerRegions.WorldTickData tickData) { // Canvas - Threaded Regions
         ChunkPos pos = chunk.getPos();
         chunk.incrementInhabitedTime(timeInhabited);
         if (true) { // Paper - rewrite chunk system
@@ -619,7 +_,7 @@
         }
 
         if (!spawnCategories.isEmpty()) {
-            if (this.level.getWorldBorder().isWithinBounds(pos)) { // Paper - rewrite chunk system
+            if (this.level.getWorldBorder().isWithinBounds(pos) && (!Config.INSTANCE.entities.enableAsyncSpawning || tickData.spawnCountsReady.get())) { // Paper - rewrite chunk system // Canvas - Threaded Regions
                 NaturalSpawner.spawnForChunk(this.level, chunk, spawnState, spawnCategories);
             }
         }
@@ -666,19 +_,26 @@
         int sectionPosX = SectionPos.blockToSectionCoord(pos.getX());
         int sectionPosZ = SectionPos.blockToSectionCoord(pos.getZ());
         ChunkHolder visibleChunkIfPresent = this.getVisibleChunkIfPresent(ChunkPos.asLong(sectionPosX, sectionPosZ));
-        if (visibleChunkIfPresent != null && visibleChunkIfPresent.blockChanged(pos)) {
-            this.chunkHoldersToBroadcast.add(visibleChunkIfPresent);
+        if (visibleChunkIfPresent != null) { visibleChunkIfPresent.blockChanged(pos); // Canvas - Threaded Regions
+            ServerRegions.getTickData(this.level).getChunkHoldersToBroadcast().add(visibleChunkIfPresent); // Canvas - Threaded Regions
         }
     }
 
     @Override
     public void onLightUpdate(LightLayer type, SectionPos pos) {
-        this.mainThreadProcessor.execute(() -> {
+        Runnable run = () -> { // Canvas - Threaded Regions
             ChunkHolder visibleChunkIfPresent = this.getVisibleChunkIfPresent(pos.chunk().toLong());
             if (visibleChunkIfPresent != null && visibleChunkIfPresent.sectionLightChanged(type, pos.y())) {
-                this.chunkHoldersToBroadcast.add(visibleChunkIfPresent);
+                ServerRegions.getTickData(this.level).getChunkHoldersToBroadcast().add(visibleChunkIfPresent);
             }
-        });
+        // Canvas start - Threaded Regions
+        };
+        if (this.level.server.isRegionized()) {
+            MinecraftServer.getThreadedServer().taskQueue.queueChunkTask(
+                this.level, pos.getX(), pos.getZ(), run
+            );
+        } else this.mainThreadProcessor.execute(run);
+        // Canvas end
     }
 
     public void addTicket(Ticket ticket, ChunkPos chunkPos) {
@@ -698,6 +_,13 @@
         return this.ticketStorage.updateChunkForced(chunkPos, add);
     }
 
+    // Canvas start - Threaded Regions
+    public boolean updateRegionTicket(ChunkPos pos, boolean add, TicketType ticketType) {
+        Ticket ticket = new Ticket(ticketType, ChunkMap.FORCED_TICKET_LEVEL);
+        return add ? this.ticketStorage.addTicket(pos.longKey, ticket) : this.ticketStorage.removeTicket(pos.longKey, ticket);
+    }
+
+    // Canvas end
     @Override
     public LongSet getForceLoadedChunks() {
         return this.ticketStorage.getForceLoadedChunks();
@@ -769,7 +_,10 @@
     @Nullable
     @VisibleForDebug
     public NaturalSpawner.SpawnState getLastSpawnState() {
-        return this.lastSpawnState;
+        // Canvas start - Threaded Regions
+        ServerRegions.WorldTickData data = ServerRegions.getTickData(this.level);
+        return data == null ? null : data.getLastSpawnState();
+        // Canvas end
     }
 
     public void deactivateTicketsOnClosing() {
@@ -778,7 +_,7 @@
 
     public void onChunkReadyToSend(ChunkHolder chunkHolder) {
         if (chunkHolder.hasChangesToBroadcast()) {
-            this.chunkHoldersToBroadcast.add(chunkHolder);
+            ServerRegions.getTickData(this.level).getChunkHoldersToBroadcast().add(chunkHolder); // Canvas - Threaded Regions
         }
     }
 
@@ -818,16 +_,77 @@
             super.doRunTask(task);
         }
 
+        // Canvas start
+        @Override
+        public void execute(final Runnable runnable) {
+            ServerLevel level = ServerChunkCache.this.level;
+            if (level.isTicking()) {
+                level.pushTask(runnable);
+            } else {
+                super.execute(runnable);
+            }
+        }
+
+        public static Set<Thread> entityOverride = new ConcurrentSet<>();
+
+        public int size() {
+            return this.pendingRunnables.size();
+        }
+
         @Override
         public boolean pollTask() {
+            return pollTask(true);
+        }
+
+        public boolean pollTask(boolean flag) {
             // Paper start - rewrite chunk system
             final ServerChunkCache serverChunkCache = ServerChunkCache.this;
-            if (serverChunkCache.runDistanceManagerUpdates()) {
+            final ServerLevel serverLevel = serverChunkCache.level;
+            Thread current = Thread.currentThread();
+            if (current == MinecraftServer.getServer().serverThread) {
+                // we don't want to process chunk tasks on main.
+                // wake the world regardless, we may actually need it.
+                serverLevel.wake();
+                // return us having no tasks, since technically we don't on this thread
+                // we return this early because we cannot have main run distance manager updates
+                // this ensures that ticket updates arent processed on main, since if it is, then
+                // we could cause parallelism issues with plugins requesting chunks that are waiting for load
+                return false;
+            }
+            if (flag && serverChunkCache.runDistanceManagerUpdates()) {
                 return true;
             } else {
-                return super.pollTask() | ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)serverChunkCache.level).moonrise$getChunkTaskScheduler().executeMainThreadTask();
+                try {
+                    entityOverride.add(current);
+                    if (serverLevel.server.isRegionized()) {
+                        // poll our own tasks.
+                        ServerRegions.WorldTickData thisData = ServerRegions.getTickData(serverLevel);
+                        if (thisData.region != null) {
+                            // we are on a region, execute our chunk task.
+                            // unlike Folia, we still can have tasks be submitted to the
+                            // main thread executor, so we run 'super' poll aswell.
+                            final BooleanSupplier poll = () -> super.pollTask() | thisData.getTaskQueueData().executeChunkTask();
+                            return tryWake(poll);
+                        }
+                    }
+                    // not regionized OR we are on global
+                    serverLevel.taskQueueRegionData.drainGlobalChunkTasks();
+                    final BooleanSupplier poll = () -> ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)serverChunkCache.level).moonrise$getChunkTaskScheduler().executeMainThreadTask() | super.pollTask();
+                    return tryWake(poll);
+                } finally {
+                    entityOverride.remove(current);
+                }
+        // Canvas end
             }
             // Paper end - rewrite chunk system
         }
+    }
+
+    private boolean tryWake(BooleanSupplier whileTrue) {
+        while (whileTrue.getAsBoolean()) {
+            // wake the world regardless, we may actually need it.
+            ServerChunkCache.this.level.wake();
+        }
+        return false; // the "while true" is now false, so we return false.
     }
 }
