--- a/net/minecraft/world/level/chunk/status/ChunkStatusTasks.java
+++ b/net/minecraft/world/level/chunk/status/ChunkStatusTasks.java
@@ -165,7 +_,7 @@
     static CompletableFuture<ChunkAccess> generateSpawn(
         WorldGenContext worldGenContext, ChunkStep step, StaticCache2D<GenerationChunkHolder> cache, ChunkAccess chunk
     ) {
-        if (!chunk.isUpgrading()) {
+        if (false) { // Canvas - remove spawn chunk step
             worldGenContext.generator().spawnOriginalMobs(new WorldGenRegion(worldGenContext.level(), cache, step, chunk));
         }
 
