--- a/net/minecraft/world/level/chunk/status/ChunkStatusTasks.java
+++ b/net/minecraft/world/level/chunk/status/ChunkStatusTasks.java
@@ -170,7 +_,7 @@
     static CompletableFuture<ChunkAccess> generateSpawn(
         WorldGenContext worldGenContext, ChunkStep step, StaticCache2D<GenerationChunkHolder> cache, ChunkAccess chunk
     ) {
-        if (!chunk.isUpgrading()) {
+        if (!io.canvasmc.canvas.Config.INSTANCE.chunks.generation.disableSpawnChunkStep && !chunk.isUpgrading()) { // Canvas - allow disabling spawn chunk step
             worldGenContext.generator().spawnOriginalMobs(new WorldGenRegion(worldGenContext.level(), cache, step, chunk));
         }
 
