--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -18,6 +_,16 @@
      * The brand id for Paper.
      */
     Key BRAND_PAPER_ID = Key.key("papermc", "paper");
+    // Canvas start - Rebrand
+    /**
+     * The brand id for Folia.
+     */
+    Key BRAND_FOLIA_ID = Key.key("papermc", "folia");
+    /**
+     * The brand id for Canvas.
+     */
+    Key BRAND_CANVAS_ID = Key.key("canvasmc", "canvas");
+    // Canvas end
 
     /**
      * Gets the {@code ServerBuildInfo}.
