--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -925,7 +_,16 @@
             BlockState neighborState,
             RandomSource random
         ) {
-            return this.getBlock().updateShape(this.asState(), level, scheduledTickAccess, pos, direction, neighborPos, neighborState, random);
+            // Canvas start - ShapeUpdateEvent
+            org.bukkit.craftbukkit.block.data.CraftBlockData provided = this.createCraftBlockData();
+            io.canvasmc.canvas.event.ShapeUpdateEvent blockShapeEvent = new io.canvasmc.canvas.event.ShapeUpdateEvent(provided);
+            boolean canceled = !blockShapeEvent.callEvent();
+            provided = (org.bukkit.craftbukkit.block.data.CraftBlockData) blockShapeEvent.getBlock();
+            if (canceled) {
+                return provided.getState();
+            }
+            return this.getBlock().updateShape(provided.getState(), level, scheduledTickAccess, pos, direction, neighborPos, neighborState, random);
+            // Canvas end
         }
 
         public boolean isPathfindable(PathComputationType type) {
