--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/executor/RadiusAwarePrioritisedExecutor.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/executor/RadiusAwarePrioritisedExecutor.java
@@ -725,5 +_,13 @@
             // TODO implement
             return this.setPriority(priority);
         }
+        // Canvas start - rewrite moonrise executor
+
+        @Override
+        public boolean equals(final Object obj) {
+            if (!(obj instanceof Task task)) return false;
+            return this.run.equals(task.run);
+        }
+        // Canvas end
     }
 }
