--- a/net/minecraft/world/entity/animal/axolotl/AxolotlAi.java
+++ b/net/minecraft/world/entity/animal/axolotl/AxolotlAi.java
@@ -142,10 +_,20 @@
         Optional<PositionTracker> memory = entity.getBrain().getMemory(MemoryModuleType.LOOK_TARGET);
         if (memory.isPresent()) {
             BlockPos blockPos = memory.get().currentBlockPosition();
+        // Canvas start - fix folia issues
+            if (!level.hasChunkAt(blockPos)) {
+                return false;
+            }
+            if (!(level instanceof ServerLevel serverLevel)) {
+                return false;
+            }
+            if (!serverLevel.getServer().isSameThread()) {
+                return false;
+            }
             return level.isWaterAt(blockPos) == entity.isInWater();
-        } else {
-            return false;
         }
+        return false;
+        // Canvas end
     }
 
     public static void updateActivity(Axolotl axolotl) {
