--- a/net/minecraft/world/entity/animal/Fox.java
+++ b/net/minecraft/world/entity/animal/Fox.java
@@ -867,6 +_,11 @@
                 return false;
             } else {
                 ServerLevel serverLevel = getServerLevel(Fox.this.level());
+                // Canvas start - Async target finding
+                if (serverLevel == null) {
+                    return false;
+                }
+                // Canvas end - Async target finding
 
                 for (EntityReference<LivingEntity> entityReference : Fox.this.getTrustedEntities().toList()) {
                     LivingEntity livingEntity = entityReference.getEntity(serverLevel, LivingEntity.class);
@@ -874,7 +_,7 @@
                         this.trustedLastHurt = livingEntity;
                         this.trustedLastHurtBy = livingEntity.getLastHurtByMob();
                         int lastHurtByMobTimestamp = livingEntity.getLastHurtByMobTimestamp();
-                        return lastHurtByMobTimestamp != this.timestamp && this.canAttack(this.trustedLastHurtBy, this.targetConditions);
+                        return lastHurtByMobTimestamp != this.timestamp && this.canAttack(this.trustedLastHurtBy, this.getTargetConditions()); // Canvas - Async target finding
                     }
                 }
 
@@ -1117,6 +_,13 @@
             Fox.this.setSitting(false);
             super.start();
         }
+
+        // Canvas start - Async target finding
+        @Override
+        protected TypeToCheck typeToCheck() {
+            return TypeToCheck.FoxEat;
+        }
+        // Canvas end - Async target finding
     }
 
     class FoxFloatGoal extends FloatGoal {
