--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -761,12 +_,6 @@
     }
 
     @Override
-    public void endMetricsRecordingTick() {
-        super.endMetricsRecordingTick();
-        this.debugSampleSubscriptionTracker.tick(this.getTickCount());
-    }
-
-    @Override
     public SampleLogger getTickTimeLogger() {
         return this.tickTimeLogger;
     }
