--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -553,7 +_,7 @@
 
                 slot.setChanged();
                 // CraftBukkit start - Make sure the client has the right slot contents
-                if (player instanceof ServerPlayer serverPlayer && slot.getMaxStackSize() != 64) {
+                if (player instanceof ServerPlayer serverPlayer && slot.getMaxStackSize() != net.minecraft.world.Container.MAX_STACK) { // Canvas - fix excess slot updates
                     serverPlayer.connection.send(new net.minecraft.network.protocol.game.ClientboundContainerSetSlotPacket(this.containerId, this.incrementStateId(), slot.index, slot.getItem()));
                     // Updating a crafting inventory makes the client reset the result slot, have to send it again
                     if (this.getBukkitView().getType() == org.bukkit.event.inventory.InventoryType.WORKBENCH || this.getBukkitView().getType() == org.bukkit.event.inventory.InventoryType.CRAFTING) {
