--- a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.entity.projectile;
 
+import io.canvasmc.canvas.Config;
 import it.unimi.dsi.fastutil.doubles.DoubleDoubleImmutablePair;
 import java.util.List;
 import java.util.OptionalInt;
@@ -135,6 +_,11 @@
                     }
                 });
             }
+            // Canvas start - Threaded Regions
+            if (this.attachedToEntity != null && !io.canvasmc.canvas.region.ServerRegions.isTickThreadFor(this.attachedToEntity)) {
+                this.attachedToEntity = null;
+            }
+            // Canvas end
 
             if (this.attachedToEntity != null) {
                 Vec3 handHoldingItemAngle;
@@ -350,4 +_,14 @@
         double d1 = entity.position().z - this.position().z;
         return DoubleDoubleImmutablePair.of(d, d1);
     }
+    // Canvas start
+
+    @Override
+    public boolean shouldBeSaved() {
+        if (Config.INSTANCE.entities.disableFireworkSaving) {
+            return false;
+        }
+        return super.shouldBeSaved();
+    }
+    // Canvas end
 }
