--- a/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -52,11 +_,23 @@
     }
 
     private void deregisterFromCurrentOwner() {
-        // Folia - region threading - we remove the registration logic, we do not need to fetch the owner
+        // Canvas start - restore vanilla ender pearl behavior
+        if (io.canvasmc.canvas.Config.INSTANCE.restoreVanillaEnderPearlBehavior) {
+            if (this.getOwner() instanceof ServerPlayer serverPlayer) {
+                serverPlayer.deregisterEnderPearl(this);
+            }
+        }
+        // Canvas end - restore vanilla ender pearl behavior
     }
 
     private void registerToCurrentOwner() {
-        // Folia - region threading - we remove the registration logic, we do not need to fetch the owner
+        // Canvas start - restore vanilla ender pearl behavior
+        if (io.canvasmc.canvas.Config.INSTANCE.restoreVanillaEnderPearlBehavior) {
+            if (this.getOwner() instanceof ServerPlayer serverPlayer) {
+                serverPlayer.registerEnderPearl(this);
+            }
+        }
+        // Canvas end - restore vanilla ender pearl behavior
     }
 
     @Nullable
@@ -274,6 +_,12 @@
             && serverPlayer.level().getGameRules().getBoolean(GameRules.RULE_ENDER_PEARLS_VANISH_ON_DEATH)) {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else {
+            // Canvas start - config for launching ender pearls outside world border
+            //noinspection SuspiciousIndentAfterControlStatement
+            if (io.canvasmc.canvas.Config.INSTANCE.discardEnderPearlsOutsideBorder && !this.level().getWorldBorder().isWithinBounds(this.position())) {
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.OUT_OF_WORLD);
+            } else
+            // Canvas end - config for launching ender pearls outside world border
             super.tick();
         }
 
