--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -89,7 +_,7 @@
     // Paper start - Allow ArmorStands not to tick
     public boolean canTick = true;
     public boolean canTickSetByAPI = false;
-    private boolean noTickEquipmentDirty = false;
+    public boolean noTickEquipmentDirty = false; // Canvas - private -> public
     // Paper end - Allow ArmorStands not to tick
     public boolean canMovementTick = true; // Purpur - Movement options for armor stands
 
@@ -528,8 +_,9 @@
         maxUpStep = level().purpurConfig.armorstandStepHeight; // Purpur - Add option to set armorstand step height
         if (!this.canTick) {
             if (this.noTickEquipmentDirty) {
-                this.noTickEquipmentDirty = false;
+                // this.noTickEquipmentDirty = false; // Canvas - equipment tracking - Remove dirty mark after handling equipment update for armor stand
                 this.detectEquipmentUpdates();
+                this.noTickEquipmentDirty = false; // Canvas - equipment tracking - Remove dirty mark after handling equipment update for armor stand
             }
 
             return;
