--- a/net/minecraft/server/commands/DebugCommand.java
+++ b/net/minecraft/server/commands/DebugCommand.java
@@ -67,6 +_,8 @@
     }
 
     private static int start(CommandSourceStack source) throws CommandSyntaxException {
+        // Canvas start - remove vanilla profiler
+        /*
         MinecraftServer server = source.getServer();
         if (server.isTimeProfilerRunning()) {
             throw ERROR_ALREADY_RUNNING.create();
@@ -75,9 +_,14 @@
             source.sendSuccess(() -> Component.translatable("commands.debug.started"), true);
             return 0;
         }
+        */
+        return 0;
+        // Canvas end
     }
 
     private static int stop(CommandSourceStack source) throws CommandSyntaxException {
+        // Canvas start - remove vanilla profiler
+        /*
         MinecraftServer server = source.getServer();
         if (!server.isTimeProfilerRunning()) {
             throw ERROR_NOT_RUNNING.create();
@@ -93,6 +_,9 @@
             );
             return (int)d1;
         }
+        */
+        return 0;
+        // Canvas end
     }
 
     static class TraceCustomExecutor
