--- a/net/minecraft/server/commands/ForceLoadCommand.java
+++ b/net/minecraft/server/commands/ForceLoadCommand.java
@@ -158,8 +_,8 @@
             int sectionPosCoord2 = SectionPos.blockToSectionCoord(max);
             int sectionPosCoord3 = SectionPos.blockToSectionCoord(max1);
             long l = (sectionPosCoord2 - sectionPosCoord + 1L) * (sectionPosCoord3 - sectionPosCoord1 + 1L);
-            if (l > 256L) {
-                throw ERROR_TOO_MANY_CHUNKS.create(256, l);
+            if (l > 25600L) { // Canvas - use higher value
+                throw ERROR_TOO_MANY_CHUNKS.create(25600, l); // Canvas - use higher value
             } else {
                 ServerLevel level = source.getLevel();
                 ResourceKey<Level> resourceKey = level.dimension();
