--- a/net/minecraft/server/commands/LocateCommand.java
+++ b/net/minecraft/server/commands/LocateCommand.java
@@ -106,6 +_,35 @@
         BlockPos blockPos = BlockPos.containing(source.getPosition());
         ServerLevel level = source.getLevel();
         Stopwatch stopwatch = Stopwatch.createStarted(Util.TICKER);
+        // Canvas start - async locator
+        if (true) {
+            BlockPos originPos = BlockPos.containing(source.getPosition());
+            io.canvasmc.canvas.chunk.structure.AsyncLocator.locate(source.getLevel(), holderSet, originPos, 100, false)
+                .thenOnRegionOrNull(representable -> {
+                    Pair<BlockPos, Holder<Structure>> pair = representable.get();
+                    stopwatch.stop();
+                    if (pair != null) {
+                        showLocateResult(
+                            source,
+                            structure,
+                            originPos,
+                            pair,
+                            "commands.locate.structure.success",
+                            false,
+                            stopwatch.elapsed()
+                        );
+                    }
+                }, () -> {
+                    source.sendFailure(
+                        Component.literal(
+                            ERROR_STRUCTURE_NOT_FOUND.create(structure.asPrintable()).getMessage()
+                        )
+                    );
+                });
+
+            return 0;
+        }
+        // Canvas end - async locator
         Pair<BlockPos, Holder<Structure>> pair = level.getChunkSource().getGenerator().findNearestMapStructure(level, holderSet, blockPos, 100, false);
         stopwatch.stop();
         if (pair == null) {
