--- a/net/minecraft/server/commands/PerfCommand.java
+++ b/net/minecraft/server/commands/PerfCommand.java
@@ -42,6 +_,9 @@
     }
 
     private static int startProfilingDedicatedServer(CommandSourceStack source) throws CommandSyntaxException {
+        // Canvas start - remove vanilla profiler
+        return removedMessage(source);
+        /*
         MinecraftServer server = source.getServer();
         if (server.isRecordingMetrics()) {
             throw ERROR_ALREADY_RUNNING.create();
@@ -52,9 +_,14 @@
             source.sendSuccess(() -> Component.translatable("commands.perf.started"), false);
             return 0;
         }
+        */
+        // Canvas end
     }
 
     private static int stopProfilingDedicatedServer(CommandSourceStack source) throws CommandSyntaxException {
+        // Canvas start - remove vanilla profiler
+        return removedMessage(source);
+        /*
         MinecraftServer server = source.getServer();
         if (!server.isRecordingMetrics()) {
             throw ERROR_NOT_RUNNING.create();
@@ -62,7 +_,21 @@
             server.finishRecordingMetrics();
             return 0;
         }
-    }
+        */
+        // Canvas end
+    }
+
+    // Canvas start - remove vanilla profiler
+    private static int removedMessage(CommandSourceStack source) {
+        net.kyori.adventure.text.minimessage.MiniMessage mm = net.kyori.adventure.text.minimessage.MiniMessage.miniMessage();
+
+        source.getSender().sendMessage(mm.deserialize("<gold>Canvas has removed Mojang's Profiler. Please use <click:suggest_command:'/spark'><grey>/spark</grey></click> instead"));
+        source.getSender().sendMessage(mm.deserialize("<gold>For more information, view its documentation at"));
+        source.getSender().sendMessage(mm.deserialize("<gold><click:open_url:'https://spark.lucko.me/docs/Command-Usage'>https://spark.lucko.me/docs/Command-Usage</click>"));
+
+        return 0;
+    }
+    // Canvas end
 
     private static void saveResults(CommandSourceStack source, Path path, MinecraftServer server) {
         String string = String.format(
