--- a/net/minecraft/server/commands/WaypointCommand.java
+++ b/net/minecraft/server/commands/WaypointCommand.java
@@ -166,6 +_,17 @@
 
     private static void mutateIcon(CommandSourceStack source, WaypointTransmitter waypoint, Consumer<Waypoint.Icon> mutator) {
         ServerLevel level = source.getLevel();
+        // Canvas start - region threading - restore waypoints
+        if (waypoint instanceof net.minecraft.world.entity.Entity entity) {
+            // run this on the waypoint transmitters thread context
+            entity.getBukkitEntity().taskScheduler.schedule((entity1) -> {
+                level.getWaypointManager().untrackWaypoint(waypoint);
+                mutator.accept(waypoint.waypointIcon());
+                level.getWaypointManager().trackWaypoint(waypoint);
+            }, null, 0L);
+            return;
+        }
+        // Canvas end
         level.getWaypointManager().untrackWaypoint(waypoint);
         mutator.accept(waypoint.waypointIcon());
         level.getWaypointManager().trackWaypoint(waypoint);
