--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/task/ChunkProgressionTask.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/task/ChunkProgressionTask.java
@@ -98,4 +_,13 @@
             ", chunk: (" + this.chunkX + "," + this.chunkZ + "), hashcode: " + System.identityHashCode(this) + ", priority: " + this.getPriority() +
             ", status: " + this.getTargetStatus().toString() + ", scheduled: " + this.isScheduled() + "}";
     }
+    // Canvas start - rewrite moonrise executor
+
+    @Override
+    public boolean equals(final Object obj) {
+        if (!(obj instanceof ChunkUpgradeGenericStatusTask task)) return false;
+        return this.chunkX == task.chunkX && this.chunkZ == task.chunkZ
+            && this.world == task.world && this.getTargetStatus() == task.getTargetStatus();
+    }
+    // Canvas end
 }
