--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/task/ChunkUpgradeGenericStatusTask.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/task/ChunkUpgradeGenericStatusTask.java
@@ -215,4 +_,13 @@
         }
         this.generateTask.raisePriority(priority);
     }
+    // Canvas start - rewrite moonrise executor
+
+    @Override
+    public boolean equals(final Object obj) {
+        if (!(obj instanceof ChunkUpgradeGenericStatusTask task)) return false;
+        return this.chunkX == task.chunkX && this.chunkZ == task.chunkZ
+            && this.toStatus == task.toStatus && this.world == task.world;
+    }
+    // Canvas end
 }
