--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -71,6 +_,7 @@
     public static void main(final OptionSet optionSet) { // CraftBukkit - replaces main(String[] args)
         io.papermc.paper.util.LogManagerShutdownThread.hook(); // Paper - Improved watchdog support
         SharedConstants.tryDetectVersion();
+        io.canvasmc.canvas.Config.init(); // Canvas - init config
         /* CraftBukkit start - Replace everything
         OptionParser optionParser = new OptionParser();
         OptionSpec<Void> optionSpec = optionParser.accepts("nogui");
