--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -69,6 +_,7 @@
     )
     @DontObfuscate
     public static void main(final OptionSet optionSet) { // CraftBukkit - replaces main(String[] args)
+        io.canvasmc.canvas.Config.init(); // Canvas - init config
         io.papermc.paper.util.LogManagerShutdownThread.hook(); // Paper - Improved watchdog support
         SharedConstants.tryDetectVersion();
         /* CraftBukkit start - Replace everything
