--- a/net/minecraft/server/ServerTickRateManager.java
+++ b/net/minecraft/server/ServerTickRateManager.java
@@ -91,10 +_,7 @@
         String string = String.format("%.2f", l == 0L ? this.millisecondsPerTick() : d / l);
         this.scheduledCurrentSprintTicks = 0L;
         this.sprintTimeSpend = 0L;
-        // Paper start - silence feedback when API requests sprint
-        if (!this.silent) this.server.createCommandSourceStack().sendSuccess(() -> Component.translatable("commands.tick.sprint.report", i, string), true);
-        this.silent = false;
-        // Paper end - silence feedback when API requests sprint
+        // Canvas - rewrite scheduler
         this.remainingSprintTicks = 0L;
         this.setFrozen(this.previousIsFrozen);
         this.server.onTickRateChanged();
