--- a/net/minecraft/world/level/pathfinder/PathTypeCache.java
+++ b/net/minecraft/world/level/pathfinder/PathTypeCache.java
@@ -24,7 +_,7 @@
     }
 
     private PathType compute(BlockGetter level, BlockPos pos, int index, long packedPos) {
-        PathType pathTypeFromState = WalkNodeEvaluator.getPathTypeFromState(level, pos);
+        PathType pathTypeFromState = WalkNodeEvaluator.canvasPathType(level, pos); // Canvas - cache path type
         this.positions[index] = packedPos;
         this.pathTypes[index] = pathTypeFromState;
         return pathTypeFromState;
