--- a/ca/spottedleaf/concurrentutil/executor/thread/PrioritisedThreadPool.java
+++ b/ca/spottedleaf/concurrentutil/executor/thread/PrioritisedThreadPool.java
@@ -12,11 +_,11 @@
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Consumer;
 
-public final class PrioritisedThreadPool {
+public class PrioritisedThreadPool { // Canvas - mutable
 
     private static final Logger LOGGER = LoggerFactory.getLogger(PrioritisedThreadPool.class);
 
-    private final Consumer<Thread> threadModifier;
+    protected final Consumer<Thread> threadModifier; // Canvas - private -> protected
     private final COWArrayList<ExecutorGroup> executors = new COWArrayList<>(ExecutorGroup.class);
     private final COWArrayList<PrioritisedThread> threads = new COWArrayList<>(PrioritisedThread.class);
     private final COWArrayList<PrioritisedThread> aliveThreads = new COWArrayList<>(PrioritisedThread.class);
