--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -1113,13 +_,6 @@
         return BuiltInRegistries.ENTITY_TYPE.getOptional(ResourceLocation.tryParse(key));
     }
 
-    // Folia start - profiler
-    public final int tickTimerId;
-    public final int inactiveTickTimerId;
-    public final int passengerTickTimerId;
-    public final int passengerInactiveTickTimerId;
-    // Folia end - profiler
-
     public EntityType(
         EntityType.EntityFactory<T> factory,
         MobCategory category,
@@ -1134,14 +_,8 @@
         int updateInterval,
         String descriptionId,
         Optional<ResourceKey<LootTable>> lootTable,
-        FeatureFlagSet requiredFeatures, // Folia start - profiler
-        String id
+        FeatureFlagSet requiredFeatures
     ) {
-        this.tickTimerId = ca.spottedleaf.leafprofiler.LProfilerRegistry.GLOBAL_REGISTRY.getOrCreateTimer("Entity Tick: " + id);
-        this.inactiveTickTimerId = ca.spottedleaf.leafprofiler.LProfilerRegistry.GLOBAL_REGISTRY.getOrCreateTimer("Inactive Entity Tick: " + id);
-        this.passengerTickTimerId = ca.spottedleaf.leafprofiler.LProfilerRegistry.GLOBAL_REGISTRY.getOrCreateTimer("Passenger Entity Tick: " + id);
-        this.passengerInactiveTickTimerId = ca.spottedleaf.leafprofiler.LProfilerRegistry.GLOBAL_REGISTRY.getOrCreateTimer("Passenger Inactive Entity Tick: " + id);
-        // Folia end - profiler
         this.factory = factory;
         this.category = category;
         this.canSpawnFarFromPlayer = canSpawnFarFromPlayer;
@@ -1702,8 +_,7 @@
                 this.updateInterval,
                 this.descriptionId.get(entityType),
                 this.lootTable.get(entityType),
-                this.requiredFeatures, // Folia - profiler
-                entityType.toString()// Folia - profiler
+                this.requiredFeatures
             );
         }
     }
