--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -69,7 +_,7 @@
  * Represents the Bukkit core, for version and Server singleton handling
  */
 public final class Bukkit {
-    private static Server server;
+    public static Server server; // Canvas - private -> public
 
     /**
      * Static class cannot be initialized.
@@ -85,6 +_,12 @@
     public static Server getServer() {
         return server;
     }
+    // Canvas start - Threaded Server
+
+    public static io.canvasmc.canvas.ThreadedBukkitServer getThreadedServer() {
+        return getServer().getThreadedServer();
+    }
+    // Canvas end
 
     /**
      * Returns the de facto plugins directory, generally used for storing plugin jars to be loaded,
