--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2752,5 +_,15 @@
      * @return TPS (5s, 15s, 1m, 5m, 15m), or null if the region doesn't exist
      */
     double @Nullable [] getRegionTPS(@NotNull World world, int chunkX, int chunkZ);
+    // Canvas start - expand region api
+
+    /**
+     * Returns the region on the current thraed, or null if we are on the global tick thread
+     * <br>
+     * This will throw if not on a tick runner
+     * @return the current region
+     */
+    @Nullable io.canvasmc.canvas.threadedregions.WorldRegionData getCurrentRegion();
+    // Canvas end
     // Folia end - region TPS API
 }
