--- a/io/papermc/paper/threadedregions/scheduler/FoliaRegionScheduler.java
+++ b/io/papermc/paper/threadedregions/scheduler/FoliaRegionScheduler.java
@@ -143,6 +_,10 @@
     }
 
     private static final class Scheduler {
+        // Canvas start - rewrite scheduler
+        private Scheduler(final io.papermc.paper.threadedregions.TickRegions.TickRegionData regionData) {}
+
+        // Canvas end - rewrite scheduler
         private static final RegionizedData.RegioniserCallback<Scheduler> REGIONISER_CALLBACK = new RegionizedData.RegioniserCallback<>() {
             @Override
             public void merge(final Scheduler from, final Scheduler into, final long fromTickOffset) {
