--- a/src/main/java/org/bukkit/craftbukkit/util/BlockStateListPopulator.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/BlockStateListPopulator.java
@@ -24,7 +_,7 @@
 public class BlockStateListPopulator extends DummyGeneratorAccess {
 
     private final LevelAccessor level;
-    private final Map<BlockPos, CapturedBlock> blocks = new LinkedHashMap<>();
+    private final Map<BlockPos, CapturedBlock> blocks = java.util.Collections.synchronizedMap(new LinkedHashMap<>()); // Canvas - synchronized
 
     private List<CraftBlockState> snapshots;
 
