--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -285,7 +_,7 @@
         }
 
         this.connection.setReadOnly();
-        this.server.scheduleOnMain(() -> {
+        io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTask(() -> { // Canvas - region threading
             this.disconnect(disconnectionInfo); // Currently you cannot cancel disconnect during the config stage
         });
     }
