--- a/net/minecraft/server/network/ServerConnectionListener.java
+++ b/net/minecraft/server/network/ServerConnectionListener.java
@@ -250,6 +_,14 @@
 
             while (iterator.hasNext()) {
                 Connection connection = iterator.next();
+                // Canvas start - connection threading
+                if (connection.packetListener != null && connection.packetListener instanceof net.minecraft.server.network.ServerConfigurationPacketListenerImpl) {
+                    // on configuration phase, move to network router
+                    connection.getPlayer().serverLevel().networkRouter.connectToWorld(connection);
+                    iterator.remove();
+                    continue;
+                }
+                // Canvas end
                 if (!connection.isConnecting()) {
                     if (connection.isConnected()) {
                         try {
@@ -268,6 +_,7 @@
                         if (connection.preparing) continue; // Spigot - Fix a race condition where a NetworkManager could be unregistered just before connection
                         iterator.remove();
                         connection.handleDisconnection();
+                        connection.removeFromAll(); // Canvas - connection threading
                     }
                 }
             }
