--- a/net/minecraft/server/network/ServerConnectionListener.java
+++ b/net/minecraft/server/network/ServerConnectionListener.java
@@ -250,6 +_,16 @@
 
             while (iterator.hasNext()) {
                 Connection connection = iterator.next();
+                // Canvas start - connection threading
+                if (connection.packetListener != null &&
+                    (connection.packetListener instanceof net.minecraft.server.network.ServerConfigurationPacketListenerImpl || connection.packetListener instanceof net.minecraft.server.network.ServerGamePacketListenerImpl)
+                ) {
+                    // on configuration phase, move to network router
+                    connection.getPlayer().serverLevel().networkRouter.connectToWorld(connection);
+                    iterator.remove();
+                    continue;
+                }
+                // Canvas end
                 if (!connection.isConnecting()) {
                     if (connection.isConnected()) {
                         try {
