--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -1,5 +_,6 @@
 package net.minecraft.server.network;
 
+import ca.spottedleaf.concurrentutil.util.Priority;
 import com.mojang.authlib.GameProfile;
 import com.mojang.logging.LogUtils;
 import java.util.List;
@@ -28,13 +_,16 @@
 import net.minecraft.server.RegistryLayer;
 import net.minecraft.server.ServerLinks;
 import net.minecraft.server.level.ClientInformation;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.server.level.TicketType;
 import net.minecraft.server.network.config.JoinWorldTask;
 import net.minecraft.server.network.config.ServerResourcePackConfigurationTask;
 import net.minecraft.server.network.config.SynchronizeRegistriesTask;
 import net.minecraft.server.packs.repository.KnownPack;
 import net.minecraft.server.players.PlayerList;
 import net.minecraft.world.flag.FeatureFlags;
+import net.minecraft.world.level.chunk.status.ChunkStatus;
 import org.slf4j.Logger;
 
 public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketListenerImpl implements ServerConfigurationPacketListener, TickablePacketListener {
@@ -153,7 +_,7 @@
                 return;
             }
 
-            Component component = org.bukkit.craftbukkit.event.CraftEventFactory.handleLoginResult(playerList.canPlayerLogin(this.connection.getRemoteAddress(), this.gameProfile), this.paperConnection, this.connection, this.gameProfile, this.server, false); // Paper - Login event logic
+            Component component = org.bukkit.craftbukkit.event.CraftEventFactory.handleLoginResult(playerList.canPlayerLogin(this.connection.getRemoteAddress(), this.gameProfile, this.connection), this.paperConnection, this.connection, this.gameProfile, this.server, false); // Paper - Login event logic // Canvas - Threaded Regions
             if (component != null) {
                 this.disconnect(component);
                 return;
@@ -169,7 +_,43 @@
                 serverPlayer = new ServerPlayer(this.server, this.server.overworld(), this.gameProfile, this.clientInformation);
             }
             // Paper end
-            playerList.placeNewPlayer(this.connection, serverPlayer, this.createCookie(this.clientInformation));
+            // Canvas start - Threaded Regions
+            final int chunkX = serverPlayer.chunkPosition().x;
+            final int chunkZ = serverPlayer.chunkPosition().z;
+            final ServerLevel world = serverPlayer.serverLevel();
+            // place ticket
+            LOGGER.info("Placing ticket for login and queuing placement task for player '{}'", serverPlayer.getName().getString());
+            Runnable placementTask = () -> {
+                try {
+                    playerList.placeNewPlayer(this.connection, serverPlayer, this.createCookie(this.clientInformation));
+                } catch (Exception e) {
+                    MinecraftServer.getThreadedServer().taskQueue.queueTickTaskQueue(
+                        world, chunkX, chunkZ,
+                        () -> {
+                            LOGGER.error("Unable to load new player, removing login ticket", e);
+                            world.getChunkSource().updateRegionTicket(serverPlayer.chunkPosition(), false, TicketType.LOGIN);
+                        }, Priority.HIGHER
+                    );
+                } finally {
+                    serverPlayer.getBukkitEntity().taskScheduler.schedule((_) -> {
+                        LOGGER.info("Finished placing new player, removing login ticket");
+                        world.getChunkSource().updateRegionTicket(serverPlayer.chunkPosition(), false, TicketType.LOGIN);
+                    }, null, 0);
+                }
+            };
+            world.loadChunksForMoveAsync(
+                serverPlayer.getBoundingBox().inflate(5),
+                Priority.HIGHER,
+                (_) -> {
+                    world.getChunkSource().updateRegionTicket(chunkX, chunkZ, true, TicketType.LOGIN);
+                    // queue tick task
+                    MinecraftServer.getThreadedServer().taskQueue.queueTickTaskQueue(
+                        world, chunkX, chunkZ,
+                        placementTask, Priority.HIGHER
+                    );
+                }
+                );
+            // Canvas end
         } catch (Exception var5) {
             LOGGER.error("Couldn't place player in world", (Throwable)var5);
             // Paper start - Debugging
@@ -177,8 +_,8 @@
                 var5.printStackTrace();
             }
             // Paper end - Debugging
-            this.connection.send(new ClientboundDisconnectPacket(DISCONNECT_REASON_INVALID_DATA));
-            this.connection.disconnect(DISCONNECT_REASON_INVALID_DATA);
+            this.connection.send(new ClientboundDisconnectPacket(Component.literal(var5.getMessage()))); // Canvas - use actual reason
+            this.connection.disconnect(Component.literal(var5.getMessage())); // Canvas - use actual reason
         }
     }
 
