--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
@@ -44,7 +_,7 @@
 
     public final Level world;
 
-    protected final SWMRLong2ObjectHashTable<ChunkSlicesRegion> regions = new SWMRLong2ObjectHashTable<>(128, 0.5f);
+    protected final it.unimi.dsi.fastutil.longs.Long2ObjectArrayMap<ChunkSlicesRegion> regions = new it.unimi.dsi.fastutil.longs.Long2ObjectArrayMap<>(128); // Canvas - optimized collection - this is actually a faster implementation than SWMRLong2ObjectHashTable
 
     protected final LevelCallback<Entity> worldCallback;
 
