--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -4,6 +_,7 @@
 import com.google.common.collect.Queues;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import com.mojang.logging.LogUtils;
+import io.canvasmc.canvas.region.ServerRegions;
 import io.netty.bootstrap.Bootstrap;
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelException;
@@ -37,7 +_,9 @@
 import java.nio.channels.ClosedChannelException;
 import java.util.Objects;
 import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
@@ -58,7 +_,9 @@
 import net.minecraft.network.protocol.login.LoginProtocols;
 import net.minecraft.network.protocol.status.ClientStatusPacketListener;
 import net.minecraft.network.protocol.status.StatusProtocols;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.RunningOnDifferentThreadException;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.util.Mth;
 import net.minecraft.util.debugchart.LocalSampleLogger;
 import org.apache.commons.lang3.Validate;
@@ -96,7 +_,7 @@
     @Nullable
     private volatile PacketListener disconnectListener;
     @Nullable
-    private volatile PacketListener packetListener;
+    public volatile PacketListener packetListener; // Canvas - private -> public - connection threading
     @Nullable
     private DisconnectionDetails disconnectionDetails;
     private boolean encrypted;
@@ -149,6 +_,27 @@
     public boolean isPending = true;
     public boolean queueImmunity;
     // Paper end - Optimize network
+    // Canvas start - packet processor
+    // if owner ref is null, it's being routed, else, it's owned by something
+    public final AtomicReference<ServerRegions.WorldTickData> owner = new AtomicReference<>();
+    public void computeIfOwningTickDataPresent(Consumer<ServerRegions.WorldTickData> ownerConsumer) {
+        if (owner.get() != null) {
+            ownerConsumer.accept(owner.get());
+        }
+    }
+    private final Queue<Runnable> packetsToProcess = new ConcurrentLinkedQueue<>();
+
+    public void schedulePacket(Runnable runnable) {
+        packetsToProcess.add(runnable);
+    }
+
+    public void tickPackets() {
+        Runnable task;
+        while ((task = packetsToProcess.poll()) != null) {
+            task.run();
+        }
+    }
+    // Canvas end
 
     public Connection(PacketFlow receiving) {
         this.receiving = receiving;
@@ -610,13 +_,7 @@
             if (!(this.packetListener instanceof net.minecraft.server.network.ServerLoginPacketListenerImpl loginPacketListener)
                 || loginPacketListener.state != net.minecraft.server.network.ServerLoginPacketListenerImpl.State.VERIFYING
                 || Connection.joinAttemptsThisTick++ < MAX_PER_TICK) {
-            // Paper start - detailed watchdog information
-            net.minecraft.network.protocol.PacketUtils.packetProcessing.push(this.packetListener);
-            try {
             tickablePacketListener.tick();
-            } finally {
-                net.minecraft.network.protocol.PacketUtils.packetProcessing.pop();
-            } // Paper end - detailed watchdog information
             } // Paper end - Buffer joins to world
         }
 
@@ -635,6 +_,7 @@
         if (this.bandwidthDebugMonitor != null) {
             this.bandwidthDebugMonitor.tick();
         }
+        this.tickPackets(); // Canvas - packet processor
     }
 
     protected void tickSecond() {
@@ -923,6 +_,19 @@
         }
         this.pendingActions.clear();
     }
+    // Canvas start - connection threading
+
+    public void removeFromAll() {
+        for (final ServerLevel level : MinecraftServer.getServer().getAllLevels()) {
+            // remove from level
+            level.levelTickData.connections.remove(this);
+            level.regioniser.computeForAllRegionsUnsynchronised((region) -> {
+                // remove from all regions
+                region.getData().tickData.connections.remove(this);
+            });
+        }
+    }
+    // Canvas end
 
     private static class InnerUtil { // Attempt to hide these methods from ProtocolLib, so it doesn't accidently pick them up.
 
