--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -4,6 +_,8 @@
 import com.google.common.collect.Queues;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import com.mojang.logging.LogUtils;
+import io.canvasmc.canvas.region.ServerRegions;
+import io.canvasmc.canvas.server.network.ConnectionHandlePhases;
 import io.netty.bootstrap.Bootstrap;
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelException;
@@ -37,7 +_,9 @@
 import java.nio.channels.ClosedChannelException;
 import java.util.Objects;
 import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
@@ -58,7 +_,9 @@
 import net.minecraft.network.protocol.login.LoginProtocols;
 import net.minecraft.network.protocol.status.ClientStatusPacketListener;
 import net.minecraft.network.protocol.status.StatusProtocols;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.RunningOnDifferentThreadException;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.util.Mth;
 import net.minecraft.util.debugchart.LocalSampleLogger;
 import org.apache.commons.lang3.Validate;
@@ -149,6 +_,49 @@
     public boolean isPending = true;
     public boolean queueImmunity;
     // Paper end - Optimize network
+    // Canvas start - packet processor
+    private final AtomicReference<ServerRegions.WorldTickData> currentlyOwned = new AtomicReference<>();
+
+    public void switchTo(ServerRegions.WorldTickData data) {
+        synchronized (lockObject) {
+            if (this.currentlyOwned.get() != null) {
+                this.currentlyOwned.get().connections.remove(this);
+            }
+            this.currentlyOwned.set(data);
+            if (data != null) data.connections.add(this);
+            this.phase = ConnectionHandlePhases.PLAY;
+        }
+    }
+
+    private final Queue<Runnable> packetsToProcess = new ConcurrentLinkedQueue<>();
+    private final Object lockObject = new Object();
+    private ConnectionHandlePhases phase = ConnectionHandlePhases.JOIN;
+
+    public void schedulePacket(Runnable runnable) {
+        packetsToProcess.add(runnable);
+    }
+
+    public void tickPackets() {
+        Runnable task;
+        while ((task = packetsToProcess.poll()) != null) {
+            task.run();
+        }
+    }
+
+    public void setPhase(ConnectionHandlePhases phase) {
+        synchronized (lockObject) {
+            this.phase = phase;
+        }
+    }
+
+    public ConnectionHandlePhases getPhase() {
+        ConnectionHandlePhases phase;
+        synchronized (lockObject) {
+            phase = this.phase;
+        }
+        return phase;
+    }
+    // Canvas end
 
     public Connection(PacketFlow receiving) {
         this.receiving = receiving;
@@ -590,6 +_,12 @@
     private static int currTick; // Paper - Buffer joins to world
     private static int tickSecond; // Purpur - Max joins per second
     public void tick() {
+        synchronized (lockObject) { // Canvas - lock
+            // Canvas start - ensure not on main with play phase
+            if (Thread.currentThread().equals(MinecraftServer.getServer().serverThread) && this.phase.equals(ConnectionHandlePhases.PLAY)) {
+                throw new IllegalStateException("Cannot run on main with play phase");
+            }
+            // Canvas end
         this.flushQueue();
         // Paper start - Buffer joins to world
         if (Connection.currTick != net.minecraft.server.MinecraftServer.currentTick) {
@@ -610,13 +_,7 @@
             if (!(this.packetListener instanceof net.minecraft.server.network.ServerLoginPacketListenerImpl loginPacketListener)
                 || loginPacketListener.state != net.minecraft.server.network.ServerLoginPacketListenerImpl.State.VERIFYING
                 || Connection.joinAttemptsThisTick++ < MAX_PER_TICK) {
-            // Paper start - detailed watchdog information
-            net.minecraft.network.protocol.PacketUtils.packetProcessing.push(this.packetListener);
-            try {
             tickablePacketListener.tick();
-            } finally {
-                net.minecraft.network.protocol.PacketUtils.packetProcessing.pop();
-            } // Paper end - detailed watchdog information
             } // Paper end - Buffer joins to world
         }
 
@@ -635,6 +_,8 @@
         if (this.bandwidthDebugMonitor != null) {
             this.bandwidthDebugMonitor.tick();
         }
+        this.tickPackets(); // Canvas - packet processor
+        } // Canvas - lock
     }
 
     protected void tickSecond() {
