--- a/net/minecraft/network/FriendlyByteBuf.java
+++ b/net/minecraft/network/FriendlyByteBuf.java
@@ -9,6 +_,7 @@
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.DynamicOps;
 import com.mojang.serialization.JsonOps;
+import io.canvasmc.canvas.Config;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufAllocator;
 import io.netty.buffer.ByteBufInputStream;
@@ -55,6 +_,7 @@
 import net.minecraft.nbt.Tag;
 import net.minecraft.network.codec.StreamDecoder;
 import net.minecraft.network.codec.StreamEncoder;
+import net.minecraft.network.protocol.status.ServerStatus;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.util.Crypt;
@@ -346,7 +_,12 @@
     }
 
     public FriendlyByteBuf writeLongArray(long[] array) {
-        writeLongArray(this, array);
+        // Canvas start - bulk write
+        VarInt.write(this, array.length);
+        ByteBuffer buf = ByteBuffer.allocate(array.length * Long.BYTES);
+        buf.asLongBuffer().put(array);
+        this.source.writeBytes(buf);
+        // Canvas end
         return this;
     }
 
