--- a/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
+++ b/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
@@ -260,7 +_,17 @@
         MinecraftServer server = MinecraftServer.getServer();
         boolean isCompleteFromCache = this.completeFromCache(true, onlineMode);
         if (onlineMode && (!isCompleteFromCache || (textures && !hasTextures()))) {
-            ProfileResult result = server.getSessionService().fetchProfile(this.profile.getId(), true);
+            // Canvas start
+            ProfileResult result;
+            try {
+                result = server.getSessionService().fetchProfile(this.profile.getId(), true);
+            } catch (com.mojang.authlib.exceptions.MinecraftClientException exception) {
+                if (exception.getMessage().contains("429")) {
+                    // too many requests, silence.
+                    result = null;
+                } else throw exception;
+            }
+            // Canvas end
             if (result != null && result.profile() != null) {
                 copyProfileProperties(result.profile(), this.profile, true);
             }
