--- a/com/mojang/math/Transformation.java
+++ b/com/mojang/math/Transformation.java
@@ -52,6 +_,7 @@
         } else {
             this.matrix = matrix;
         }
+        this.ensureDecomposed(); // Canvas - rewrite moonrise executor
     }
 
     public Transformation(@Nullable Vector3f translation, @Nullable Quaternionf leftRotation, @Nullable Vector3f scale, @Nullable Quaternionf rightRotation) {
@@ -61,6 +_,7 @@
         this.scale = scale != null ? scale : new Vector3f(1.0F, 1.0F, 1.0F);
         this.rightRotation = rightRotation != null ? rightRotation : new Quaternionf();
         this.decomposed = true;
+        this.ensureDecomposed(); // Canvas - rewrite moonrise executor
     }
 
     public static Transformation identity() {
