--- a/io/papermc/paper/redstone/RedstoneWireTurbo.java
+++ b/io/papermc/paper/redstone/RedstoneWireTurbo.java
@@ -281,7 +_,7 @@
      * Keep track of all block positions discovered during search and their current states.
      * We want to remember one entry for each position.
      */
-    private final Map<BlockPos, UpdateNode> nodeCache = Maps.newHashMap();
+    private final Map<BlockPos, UpdateNode> nodeCache = new it.unimi.dsi.fastutil.objects.Reference2ReferenceOpenHashMap<>(); // Canvas - optimized collection
 
     /*
      * For a newly created UpdateNode object, determine what type of block it is.
@@ -829,14 +_,14 @@
         j = getMaxCurrentStrength(upd, j);
         int l = 0;
 
-        wire.shouldSignal = false;
+        wire.shouldSignal.set(false); // Canvas - threadlocal
         // Unfortunately, World.isBlockIndirectlyGettingPowered is complicated,
         // and I'm not ready to try to replicate even more functionality from
         // elsewhere in Minecraft into this accelerator.  So sadly, we must
         // suffer the performance hit of this very expensive call.  If there
         // is consistency to what this call returns, we may be able to cache it.
         final int k = worldIn.getBestNeighborSignal(upd.self);
-        wire.shouldSignal = true;
+        wire.shouldSignal.set(true); // Canvas - threadlocal
 
         // The variable 'k' holds the maximum redstone power value of any adjacent blocks.
         // If 'k' has the highest level of all neighbors, then the power level of this
