--- a/net/minecraft/stats/StatType.java
+++ b/net/minecraft/stats/StatType.java
@@ -11,7 +_,7 @@
 
 public class StatType<T> implements Iterable<Stat<T>> {
     private final Registry<T> registry;
-    private final Map<T, Stat<T>> map = new IdentityHashMap<>();
+    private final Map<T, Stat<T>> map = new it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap<>(); // Canvas - optimized collection
     private final Component displayName;
     private final StreamCodec<RegistryFriendlyByteBuf, Stat<T>> streamCodec;
 
