--- a/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidPiece.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidPiece.java
@@ -23,29 +_,44 @@
 public class DesertPyramidPiece extends ScatteredFeaturePiece {
     public static final int WIDTH = 21;
     public static final int DEPTH = 21;
-    private final boolean[] hasPlacedChest = new boolean[4];
-    private final List<BlockPos> potentialSuspiciousSandWorldPositions = new ArrayList<>();
+    private final java.util.concurrent.atomic.AtomicReferenceArray<Boolean> hasPlacedChestAtomic = new java.util.concurrent.atomic.AtomicReferenceArray<>(new Boolean[4]); // Canvas - rewrite moonrise executor - fix threading issues - atomic
+    private final java.util.Set<BlockPos> potentialSuspiciousSandWorldPositions = com.google.common.collect.Sets.newConcurrentHashSet(); // Canvas - rewrite moonrise executor - fix threading issues - concurrent
     private BlockPos randomCollapsedRoofPos = BlockPos.ZERO;
+    // Canvas start - rewrite moonrise executor - fix threading issues
+    private void onInit() {
+        for(int i = 0; i < this.hasPlacedChestAtomic.length(); ++i) {
+            if (this.hasPlacedChestAtomic.get(i) == null) {
+                this.hasPlacedChestAtomic.set(i, false);
+            }
+        }
+    }
+    // Canvas end
 
     public DesertPyramidPiece(RandomSource random, int x, int z) {
         super(StructurePieceType.DESERT_PYRAMID_PIECE, x, 64, z, 21, 15, 21, getRandomHorizontalDirection(random));
+        onInit(); // Canvas - rewrite moonrise executor - fix threading issues
     }
 
     public DesertPyramidPiece(CompoundTag tag) {
         super(StructurePieceType.DESERT_PYRAMID_PIECE, tag);
-        this.hasPlacedChest[0] = tag.getBooleanOr("hasPlacedChest0", false);
-        this.hasPlacedChest[1] = tag.getBooleanOr("hasPlacedChest1", false);
-        this.hasPlacedChest[2] = tag.getBooleanOr("hasPlacedChest2", false);
-        this.hasPlacedChest[3] = tag.getBooleanOr("hasPlacedChest3", false);
+        // Canvas start - rewrite moonrise executor - fix threading issues
+        this.hasPlacedChestAtomic.set(0, tag.getBooleanOr("hasPlacedChest0", false));
+        this.hasPlacedChestAtomic.set(1, tag.getBooleanOr("hasPlacedChest1", false));
+        this.hasPlacedChestAtomic.set(2, tag.getBooleanOr("hasPlacedChest2", false));
+        this.hasPlacedChestAtomic.set(3, tag.getBooleanOr("hasPlacedChest3", false));
+        onInit();
+        // Canvas end
     }
 
     @Override
     protected void addAdditionalSaveData(StructurePieceSerializationContext context, CompoundTag tag) {
         super.addAdditionalSaveData(context, tag);
-        tag.putBoolean("hasPlacedChest0", this.hasPlacedChest[0]);
-        tag.putBoolean("hasPlacedChest1", this.hasPlacedChest[1]);
-        tag.putBoolean("hasPlacedChest2", this.hasPlacedChest[2]);
-        tag.putBoolean("hasPlacedChest3", this.hasPlacedChest[3]);
+        // Canvas start - rewrite moonrise executor - fix threading issues
+        tag.putBoolean("hasPlacedChest0", this.hasPlacedChestAtomic.get(0));
+        tag.putBoolean("hasPlacedChest1", this.hasPlacedChestAtomic.get(1));
+        tag.putBoolean("hasPlacedChest2", this.hasPlacedChestAtomic.get(2));
+        tag.putBoolean("hasPlacedChest3", this.hasPlacedChestAtomic.get(3));
+        // Canvas end
     }
 
     @Override
@@ -287,12 +_,12 @@
             this.placeBlock(level, Blocks.CUT_SANDSTONE.defaultBlockState(), 10, -11, 13, box);
 
             for (Direction direction : Direction.Plane.HORIZONTAL) {
-                if (!this.hasPlacedChest[direction.get2DDataValue()]) {
+                if (!this.hasPlacedChestAtomic.get(direction.get2DDataValue())) { // Canvas - rewrite moonrise executor - fix threading issues
                     int i4 = direction.getStepX() * 2;
                     int i5 = direction.getStepZ() * 2;
-                    this.hasPlacedChest[direction.get2DDataValue()] = this.createChest(
+                    this.hasPlacedChestAtomic.set(direction.get2DDataValue(), this.createChest( // Canvas - rewrite moonrise executor - fix threading issues
                         level, box, random, 10 + i4, -11, 10 + i5, BuiltInLootTables.DESERT_PYRAMID
-                    );
+                    )); // Canvas - rewrite moonrise executor - fix threading issues
                 }
             }
 
@@ -419,7 +_,7 @@
         this.randomCollapsedRoofPos = new BlockPos(this.getWorldX(i1, randomInt), this.getWorldY(y), this.getWorldZ(i1, randomInt));
     }
 
-    public List<BlockPos> getPotentialSuspiciousSandWorldPositions() {
+    public java.util.Set<BlockPos> getPotentialSuspiciousSandWorldPositions() { // Canvas - rewrite moonrise executor - fix threading issues
         return this.potentialSuspiciousSandWorldPositions;
     }
 
