--- a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
@@ -91,7 +_,7 @@
     public static class MineShaftCorridor extends MineshaftPieces.MineShaftPiece {
         private final boolean hasRails;
         private final boolean spiderCorridor;
-        private boolean hasPlacedSpider;
+        private volatile boolean hasPlacedSpider; // Canvas - rewrite moonrise executor - fix threading issues - volatile
         private final int numSections;
 
         public MineShaftCorridor(CompoundTag tag) {
@@ -950,7 +_,7 @@
     }
 
     public static class MineShaftRoom extends MineshaftPieces.MineShaftPiece {
-        private final List<BoundingBox> childEntranceBoxes = Lists.newLinkedList();
+        private final List<BoundingBox> childEntranceBoxes = java.util.Collections.synchronizedList(Lists.newLinkedList()); // Canvas - rewrite moonrise executor - fix threading issues - synchronized
 
         public MineShaftRoom(int genDepth, RandomSource random, int x, int z, MineshaftStructure.Type type) {
             super(
