--- a/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java
@@ -126,7 +_,7 @@
             int i = 11;
             this.entranceX = 7;
             this.entranceY = 4;
-            this.baseGrid = new WoodlandMansionPieces.SimpleGrid(11, 11, 5);
+            this.baseGrid = new io.canvasmc.canvas.util.chunk.ConcurrentFlagMatrix(11, 11, 5); // Canvas - rewrite moonrise executor - fix threading issues - concurrent
             this.baseGrid.set(this.entranceX, this.entranceY, this.entranceX + 1, this.entranceY + 1, 3);
             this.baseGrid.set(this.entranceX - 1, this.entranceY, this.entranceX - 1, this.entranceY + 1, 2);
             this.baseGrid.set(this.entranceX + 2, this.entranceY - 2, this.entranceX + 3, this.entranceY + 3, 5);
@@ -145,14 +_,14 @@
             }
 
             this.floorRooms = new WoodlandMansionPieces.SimpleGrid[3];
-            this.floorRooms[0] = new WoodlandMansionPieces.SimpleGrid(11, 11, 5);
-            this.floorRooms[1] = new WoodlandMansionPieces.SimpleGrid(11, 11, 5);
-            this.floorRooms[2] = new WoodlandMansionPieces.SimpleGrid(11, 11, 5);
+            this.floorRooms[0] = new io.canvasmc.canvas.util.chunk.ConcurrentFlagMatrix(11, 11, 5); // Canvas - rewrite moonrise executor - fix threading issues - concurrent
+            this.floorRooms[1] = new io.canvasmc.canvas.util.chunk.ConcurrentFlagMatrix(11, 11, 5); // Canvas - rewrite moonrise executor - fix threading issues - concurrent
+            this.floorRooms[2] = new io.canvasmc.canvas.util.chunk.ConcurrentFlagMatrix(11, 11, 5); // Canvas - rewrite moonrise executor - fix threading issues - concurrent
             this.identifyRooms(this.baseGrid, this.floorRooms[0]);
             this.identifyRooms(this.baseGrid, this.floorRooms[1]);
             this.floorRooms[0].set(this.entranceX + 1, this.entranceY, this.entranceX + 1, this.entranceY + 1, 8388608);
             this.floorRooms[1].set(this.entranceX + 1, this.entranceY, this.entranceX + 1, this.entranceY + 1, 8388608);
-            this.thirdFloorGrid = new WoodlandMansionPieces.SimpleGrid(this.baseGrid.width, this.baseGrid.height, 5);
+            this.thirdFloorGrid = new io.canvasmc.canvas.util.chunk.ConcurrentFlagMatrix(this.baseGrid.width, this.baseGrid.height, 5); // Canvas - rewrite moonrise executor - fix threading issues - concurrent
             this.setupThirdFloor();
             this.identifyRooms(this.thirdFloorGrid, this.floorRooms[2]);
         }
@@ -1139,9 +_,9 @@
     }
 
     static class PlacementData {
-        public Rotation rotation;
-        public BlockPos position;
-        public String wallType;
+        public volatile Rotation rotation; // Canvas - rewrite moonrise executor - fix threading issues - volatile
+        public volatile BlockPos position; // Canvas - rewrite moonrise executor - fix threading issues - volatile
+        public volatile String wallType; // Canvas - rewrite moonrise executor - fix threading issues - volatile
     }
 
     static class SecondFloorRoomCollection extends WoodlandMansionPieces.FloorRoomCollection {
@@ -1181,7 +_,7 @@
         }
     }
 
-    static class SimpleGrid {
+    public static class SimpleGrid { // Canvas - rewrite moonrise executor - fix threading issues - public
         private final int[][] grid;
         final int width;
         final int height;
