--- a/com/mojang/datafixers/types/Type.java
+++ b/com/mojang/datafixers/types/Type.java
@@ -192,7 +_,7 @@
             PENDING_REWRITE_CACHE.remove(key);
             return result;
         }
-        return (Optional<RewriteResult<A, ?>>) pending.join();
+        return (Optional<RewriteResult<A, ?>>) io.canvasmc.canvas.util.CFUtil.join(pending); // Canvas - rewrite moonrise executor
     }
 
     public <FT, FR> Type<?> getSetType(final OpticFinder<FT> optic, final Type<FR> newType) {
